### ISSUE TITLE
HTML export with code requires running things twice.

### ISSUE THREAD

**Issue by patrick-kidger** [marimo-team/marimo#2284]
### Description

A notebook may contain the results of a moderately-long-running computation. (E.g. training a machine learning model for a few hours on a GPU.) Typically this was created interactively during the runtime of a `marimo edit some_notebook.py`.

However there does not appear to be any way to export the computation to HTML except via the CLI `marimo export html --include-code some_notebook.py -o some_notebook.html` -- which runs the computation again!

Unfortunately I believe this makes Marimo unsuitable as a Jupyter replacement for many ML use-cases.

### Suggested solution

I'm opening this as  a feature request as I assume this is just a missing element in the web UI. So far as I can tell `--include-code` exists only in the CLI.

### Additional context

`marimo env`:
```
{
  "marimo": "0.8.14",
  "OS": "Darwin",
  "OS Version": "23.6.0",
  "Processor": "arm",
  "Python Version": "3.11.9",
  "Binaries": {
    "Browser": "128.0.6613.120",
    "Node": "v21.6.2"
  },
  "Dependencies": {
    "click": "8.1.7",
    "importlib-resources": "missing",
    "jedi": "0.19.1",
    "markdown": "3.7",
    "pygments": "2.18.0",
    "pymdown-extensions": "10.9",
    "ruff": "0.6.4",
    "starlette": "0.38.5",
    "tomlkit": "0.13.2",
    "typing-extensions": "4.12.2",
    "uvicorn": "0.30.6",
    "websockets": "12.0"
  },
  "Optional Dependencies": {
    "pandas": "2.2.2",
    "polars": "1.6.0",
    "pyarrow": "17.0.0"
  }
}
```

**Comment by mscolnick** [comment]
Oh sorry, I misread the issue. Correct, it doesn't include-code. We can make this available in the UI

**Comment by mscolnick** [comment]
@patrick-kidger, it is called `Download as HTML`, which `--include-code` is always true. Is this what you were expecting/looking for?

**Comment by patrick-kidger** [comment]
Ah, I see! so `Toggle app view` is equivalent to `marimo export html` and `≡ > Download > Download as HTML` is equivalent to `marimo export html --include-code`.

Thanks for your help; I didn't realise that. In that case I'll close this issue.

In searching around on this issue in the mean time, I'll note in passing that:
- Doing so automatically still requires the CLI, so as compared to Jupyter then I think there's still a footgun here. (#718)
- The exported HTML does not work at all in my browser (qutebrowser). I get a blank page whilst the console produces the following output:
  ```
   Failed to load resource: net::ERR_NETWORK_ACCESS_DENIED
  noise-60BoTA8O.png:1 
   Failed to load resource: net::ERR_NETWORK_ACCESS_DENIED
  Lora-VariableFont_wght-B2ootaw-.ttf:1 
   Failed to load resource: net::ERR_NETWORK_ACCESS_DENIED
  PTSans-Regular-CxL0S8W7.ttf:1 
   Failed to load resource: net::ERR_NETWORK_ACCESS_DENIED
  PTSans-Bold-D9fedIX3.ttf:1 
   Failed to load resource: net::ERR_NETWORK_ACCESS_DENIED
  FiraMono-Regular-BTCkDNvf.ttf:1 
   Failed to load resource: net::ERR_NETWORK_ACCESS_DENIED
  FiraMono-Medium-DU3aDxX5.ttf:1 
   Failed to load resource: net::ERR_NETWORK_ACCESS_DENIED
  FiraMono-Bold-CLVRCuM9.ttf:1 
   Failed to load resource: net::ERR_NETWORK_ACCESS_DENIED
  index-ecVrZj8d.js:1 
   Failed to load resource: net::ERR_NETWORK_ACCESS_DENIED
  index-Bta_AV3w.css:1 
   Failed to load resource: net::ERR_NETWORK_ACCESS_DENIED
  cdn.jsdelivr.net/npm…/dist/favicon.ico:1 
   Failed to load resource: net::ERR_NETWORK_ACCESS_DENIED
  The resource <URL> was preloaded using link preload but not used within a few seconds from the window's load event. Please make sure it has an appropriate `as` value and it is preloaded intentionally.
  ```
  It looks like the local HTML page is trying to access remote content, [which by default is blocked by qutebrowser](https://github.com/qutebrowser/qutebrowser/issues/7842#issuecomment-1685377409). So heads-up that the current HTML format isn't portable / probably won't work offline.

**Comment by mscolnick** [comment]
Thanks for the feedback 

Can you elaborate on your first point: "Doing so automatically still requires the CLI"?

Yes, export to html currently does require loading contents from a CDN. We could add an option to inline js and css (which gets quite expensive and will bloat the html, but doable). There is an open ticket https://github.com/marimo-team/marimo/issues/575

**Comment by patrick-kidger** [comment]
> Can you elaborate on your first point: "Doing so automatically still requires the CLI"?

Sure :) That it'd be nice to have a way to autoexport a notebook to HTML, whilst editing the notebook it in the UI. Right now if the notebook holds the output of a long-running program, then it's very easy to accidentally close the notebook without exporting its results.

At the moment I believe the only way to export automatically is via `marimo export html --watch` from the CLI, but thats brings us back to the start of this issue: we end up running the notebook twice. (Once in the web UI and once from the CLI.)

---

On inlining: is all the JS really needed? At least for the 'display some code with markdown' use case (#2283, #2287), then static HTML+CSS page is probably all that is required a substantial portion of the time.
